## Chapter 5 – Exercise 1: Sử dụng ibm_db API trong Python để kết nối đến IBM DB2 warehouse

#### Học viên thực hiện các yêu cầu sau và đối chiếu với kết quả cho trước:

In [1]:
import pandas as pd
import ibm_db_dbi
import ibm_db

In [2]:
dsn_database = "BLUDB"            
dsn_hostname = "dashdb-txn-sbox-yp-dal09-08.services.dal.bluemix.net" 
dsn_port = "50000"                
dsn_protocol = "TCPIP"            
dsn_uid = "rbq33661"        
dsn_pwd = "1f2h-rm6r64019m7"

In [3]:
# Câu 1: Tạo kết nối đến Database QuanLySinhVien đã được tạo trong Chapter2_Exercise2
dsn = (
        "DATABASE = {0};"
        "HOSTNAME = {1};"
        "PORT = {2};"
        "PROTOCOL = {3};"
        "UID = {4};"
        "PWD = {5};").format(dsn_database, dsn_hostname, dsn_port, dsn_protocol, dsn_uid, dsn_pwd)
try:
    conn = ibm_db.connect("DATABASE=BLUDB;HOSTNAME=dashdb-txn-sbox-yp-dal09-08.services.dal.bluemix.net;PORT=50000;PROTOCOL=TCPIP;UID=rbq33661;PWD=1f2h-rm6r64019m7;", "", "")
    pconn = ibm_db_dbi.Connection(conn)
    print("Connected!")
except:
    print("Unable to connect to database")

Connected!


In [4]:
dsn = (
        "DATABASE = {0};"
        "HOSTNAME = {1};"
        "PORT = {2};"
        "PROTOCOL = {3};"
        "UID = {4};"
        "PWD = {5};").format(dsn_database, dsn_hostname, dsn_port, dsn_protocol, dsn_uid, dsn_pwd)
print(dsn)

DATABASE = BLUDB;HOSTNAME = dashdb-txn-sbox-yp-dal09-08.services.dal.bluemix.net;PORT = 50000;PROTOCOL = TCPIP;UID = rbq33661;PWD = 1f2h-rm6r64019m7;


In [5]:
# Câu 2: Sử dụng ibm_db API, truy vấn dữ liệu trên 1 bảng

In [6]:
def connect(dsn):
#     code here
    
    

SyntaxError: unexpected EOF while parsing (<ipython-input-6-5f965bde84d4>, line 4)

In [7]:
# Câu 2a: Liệt kê danh sách sinh viên gồm các cột Mã, Họ và Tên, Ngày sinh, Giới tính, Học bổng. Có sắp tăng theo cột học bổng.
q = '''select "Ma_sinh_vien", concat(concat ("Ho" ,\' \'),"Ten") as "Ho_ten", 
        "Gioi_tinh", "Ngay_sinh", "Hoc_bong" from sinh_vien 
        order by "Hoc_bong" desc limit 5 
    '''
df = pd.read_sql(q, pconn)
df

Ma_sinh_vien              Ho_ten Gioi_tinh  Ngay_sinh  Hoc_bong
0        C0012       Mã Vĩnh Thiệu      True 1999-10-09   3000000
1        C0033   Mã Văn Minh Trọng      True 1999-09-03   3000000
2        C0031        Lý Thảo Thúy     False 1999-07-05   3000000
3        C0025  Phạm Phương Phương     False 1999-06-15   3000000
4        C0021      Vương Vĩnh Hòa      True 1999-01-03   3000000

In [49]:
q = '''select "Ma_sinh_vien", concat(concat ("Ho" ,\' \'),"Ten") as "Ho_ten", 
        "Gioi_tinh", "Ngay_sinh", "Hoc_bong" from sinh_vien 
        order by "Hoc_bong" limit 5
    '''
df = pd.read_sql(q, pconn)
df

Ma_sinh_vien                Ho_ten Gioi_tinh  Ngay_sinh  Hoc_bong
0        C0002        Đặng Bạch Ngọc      True 1999-11-11         0
1        C0026         Lâm Vĩnh Hiệu      True 1999-07-26         0
2        C0014         Khương Thị Ly     False 1999-02-11         0
3        C0009  Thái Thị Thanh Nhung     False 1999-07-18         0
4        C0003   Phạm Văn Minh Thiện      True 1999-11-23         0

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
   <pre> 5 dòng đầu của dữ liệu:</pre>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_sinh_vien</th>
      <th>Ho_ten</th>
      <th>Gioi_tinh</th>
      <th>Ngay_sinh</th>
      <th>Hoc_bong_</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>T0507</td>
      <td>Cao Minh Thu</td>
      <td>Nữ</td>
      <td>04/05/1999 00:00:00</td>
      <td>.00</td>
    </tr>
    <tr>
      <th>1</th>
      <td>T0509</td>
      <td>Văn Phú Minh</td>
      <td>Nam</td>
      <td>26/02/1999 00:00:00</td>
      <td>.00</td>
    </tr>
    <tr>
      <th>2</th>
      <td>T0510</td>
      <td>Khương Vĩnh Dũng</td>
      <td>Nam</td>
      <td>17/04/1999 00:00:00</td>
      <td>.00</td>
    </tr>
    <tr>
      <th>3</th>
      <td>T0512</td>
      <td>Đàm Vĩnh Huy</td>
      <td>Nam</td>
      <td>18/03/1999 00:00:00</td>
      <td>.00</td>
    </tr>
    <tr>
      <th>4</th>
      <td>T0520</td>
      <td>Bạch Văn Minh Lộc</td>
      <td>Nam</td>
      <td>19/09/1999 00:00:00</td>
      <td>.00</td>
    </tr>
  </tbody>
</table>
</div>

<pre> 5 dòng cuối của dữ liệu:</pre>
<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_sinh_vien</th>
      <th>Ho_ten</th>
      <th>Gioi_tinh</th>
      <th>Ngay_sinh</th>
      <th>Hoc_bong_</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>730</th>
      <td>C0461</td>
      <td>Hoàng Thảo Hạnh</td>
      <td>Nữ</td>
      <td>21/02/1999 00:00:00</td>
      <td>3,307,500.00</td>
    </tr>
    <tr>
      <th>731</th>
      <td>C0466</td>
      <td>Lâm Thảo Hoa</td>
      <td>Nữ</td>
      <td>18/04/1999 00:00:00</td>
      <td>3,307,500.00</td>
    </tr>
    <tr>
      <th>732</th>
      <td>C0476</td>
      <td>Cao Thị Thảo Trang</td>
      <td>Nữ</td>
      <td>04/01/1999 00:00:00</td>
      <td>3,307,500.00</td>
    </tr>
    <tr>
      <th>733</th>
      <td>C0481</td>
      <td>Bạch Minh Thu</td>
      <td>Nữ</td>
      <td>24/02/1999 00:00:00</td>
      <td>3,307,500.00</td>
    </tr>
    <tr>
      <th>734</th>
      <td>C0493</td>
      <td>Đinh Phú Mỹ</td>
      <td>Nữ</td>
      <td>06/02/1999 00:00:00</td>
      <td>3,307,500.00</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [51]:
# Câu 2b: Hiển thị danh sách 10 sinh viên đầu tiên có học bổng từ cao đến thấp 
# Hướng dẫn: dùng FETCH FIRST 10 ROWS ONLY và ORDER BY … DESC
q = '''select * 
         from sinh_vien
         order by "Hoc_bong" desc
         FETCH FIRST 10 ROWS ONLY
    '''
df = pd.read_sql(q, pconn)
df

Ma_sinh_vien              Ho     Ten Gioi_tinh  Ngay_sinh  \
0        C0012         Mã Vĩnh   Thiệu      True 1999-10-09   
1        C0049          Lý Văn    Dũng      True 1999-09-12   
2        C0045        Thái Nam     Tân      True 1999-01-03   
3        C0043  Bạch Thị Thanh   Nhung     False 1999-10-30   
4        C0042        Đàm Vĩnh    Dũng      True 1999-08-04   
5        C0034       Thái Minh  Phương      True 1999-07-09   
6        C0033     Mã Văn Minh   Trọng      True 1999-09-03   
7        C0031         Lý Thảo    Thúy     False 1999-07-05   
8        C0025     Phạm Phương  Phương     False 1999-06-15   
9        C0021      Vương Vĩnh     Hòa      True 1999-01-03   

                 Mail     Di_dong          CMND  Hoc_bong Ma_khoa  
0   mvthieu@gmail.com  0168135829  522178884206   3000000      CN  
1    lvdung@gmail.com  0913541297  618232837549   3000000      CN  
2     tntan@gmail.com  0168841324  676458978939   3000000      CN  
3  bttnhung@gmail.com  0168114496   40440424749   3000000      CN  
4    dvdung@gmail.com  0913702776  237164154244   3000000      CN  
5  tmphuong@gmail.com  0987687468  925518195176   3000000      CN  
6  mvmtrong@gmail.com  0987078641  532328788440   3000000      CN  
7    ltthuy@gmail.com  0987566402   13098347676   3000000      CN  
8  ppphuong@gmail.com  0168374535  774375787574   3000000      CN  
9     vvhoa@gmail.com  0987334366  661854084622   3000000      CN

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_sinh_vien</th>
      <th>Ho</th>
      <th>Ten</th>
      <th>Gioi_tinh</th>
      <th>Ngay_sinh</th>
      <th>Mail</th>
      <th>Di_dong</th>
      <th>CMND</th>
      <th>Hoc_bong</th>
      <th>Ma_khoa</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>C0025</td>
      <td>Phạm Phương</td>
      <td>Phương</td>
      <td>False</td>
      <td>1999-06-15</td>
      <td>ppphuong@gmail.com</td>
      <td>0168374535</td>
      <td>774375787574</td>
      <td>3307500</td>
      <td>CN</td>
    </tr>
    <tr>
      <th>1</th>
      <td>C0080</td>
      <td>Bạch Thanh</td>
      <td>Thanh</td>
      <td>False</td>
      <td>1999-01-04</td>
      <td>btthanh@gmail.com</td>
      <td>0913843497</td>
      <td>169614465950</td>
      <td>3307500</td>
      <td>CN</td>
    </tr>
    <tr>
      <th>2</th>
      <td>C0075</td>
      <td>Lê Thảo</td>
      <td>Hoa</td>
      <td>False</td>
      <td>1999-05-15</td>
      <td>lthoa@gmail.com</td>
      <td>0913243932</td>
      <td>662273790256</td>
      <td>3307500</td>
      <td>CN</td>
    </tr>
    <tr>
      <th>3</th>
      <td>C0074</td>
      <td>Thái Thảo</td>
      <td>Nga</td>
      <td>False</td>
      <td>1999-02-23</td>
      <td>ttnga@gmail.com</td>
      <td>0168122579</td>
      <td>443293205719</td>
      <td>3307500</td>
      <td>CN</td>
    </tr>
    <tr>
      <th>4</th>
      <td>C0065</td>
      <td>Lâm Thảo</td>
      <td>Trúc</td>
      <td>False</td>
      <td>1999-09-19</td>
      <td>lttruc@gmail.com</td>
      <td>0987113162</td>
      <td>613542186162</td>
      <td>3307500</td>
      <td>CN</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [52]:
# Câu 3: Sử dụng ibm_db API, truy vấn trên nhiều bảng

In [58]:
# Câu 3a: Liệt kê danh sách sinh viên gồm các cột mã, họ, tên, ngày sinh, giới tính, mã khoa, tên khoa
q = '''select "Ma_sinh_vien", "Ho", s."Ten", "Ngay_sinh", "Gioi_tinh", s."Ma_khoa", k."Ten" as "Ten_khoa"
        from sinh_vien s left join khoa k on k."Ma_khoa"= s."Ma_khoa"
        '''
df = pd.read_sql(q, pconn)
df

Ma_sinh_vien              Ho    Ten  Ngay_sinh Gioi_tinh Ma_khoa  \
0          C0240        Phạm Thị  Hương 1999-06-19     False      CN   
1          C0001     Khương Thảo   Loan 1999-01-04     False      CN   
2          C0002       Đặng Bạch   Ngọc 1999-11-11      True      CN   
3          C0003   Phạm Văn Minh  Thiện 1999-11-23      True      CN   
4          C0004  Đinh Thị Thanh   Dung 1999-01-19     False      CN   
..           ...             ...    ...        ...       ...     ...   
995        V0996       Trần Mạnh  Thăng 1999-06-21      True      VL   
996        V0997       Trần Vĩnh  Thiện 1999-07-23      True      VL   
997        V0998     Khương Thảo   Hạnh 1999-08-16     False      VL   
998        V0999       Đinh Ngọc   Ngọc 1999-05-25      True      VL   
999        V1000         Lý Mạnh    Bảo 1999-06-03      True      VL   

                     Ten_khoa  
0    Khoa Công nghệ Thông tin  
1    Khoa Công nghệ Thông tin  
2    Khoa Công nghệ Thông tin  
3    Khoa Công nghệ Thông tin  
4    Khoa Công nghệ Thông tin  
..                        ...  
995               Khoa Vật lý  
996               Khoa Vật lý  
997               Khoa Vật lý  
998               Khoa Vật lý  
999               Khoa Vật lý  

[1000 rows x 7 columns]

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_sinh_vien</th>
      <th>Ho</th>
      <th>Ten</th>
      <th>Ngay_sinh</th>
      <th>Gioi_tinh</th>
      <th>Ma_khoa</th>
      <th>Ten_khoa</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>T0504</td>
      <td>Đặng Thị Thanh</td>
      <td>Lan</td>
      <td>1999-02-22</td>
      <td>False</td>
      <td>TO</td>
      <td>Khoa Toán</td>
    </tr>
    <tr>
      <th>1</th>
      <td>C0004</td>
      <td>Đinh Thị Thanh</td>
      <td>Dung</td>
      <td>1999-01-19</td>
      <td>False</td>
      <td>CN</td>
      <td>Khoa Công nghệ Thông tin</td>
    </tr>
    <tr>
      <th>2</th>
      <td>C0007</td>
      <td>Lê Thị</td>
      <td>Mai</td>
      <td>1999-09-13</td>
      <td>False</td>
      <td>CN</td>
      <td>Khoa Công nghệ Thông tin</td>
    </tr>
    <tr>
      <th>3</th>
      <td>C0009</td>
      <td>Thái Thị Thanh</td>
      <td>Nhung</td>
      <td>1999-07-18</td>
      <td>False</td>
      <td>CN</td>
      <td>Khoa Công nghệ Thông tin</td>
    </tr>
    <tr>
      <th>4</th>
      <td>C0011</td>
      <td>Hoàng Thảo</td>
      <td>Thúy</td>
      <td>1999-12-07</td>
      <td>False</td>
      <td>CN</td>
      <td>Khoa Công nghệ Thông tin</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [61]:
# Câu 3b: Liệt kê kết quả học tập gồm các cột mã sinh viên, họ và tên, mã môn học, điểm
q = '''select kq."Ma_sinh_vien", concat(concat(s."Ho", \' \'), s."Ten") as "Ho_ten", kq."Ma_mon", kq."Diem"
         from ket_qua_2 kq left join sinh_vien s on s."Ma_sinh_vien" = kq."Ma_sinh_vien"
        '''
df = pd.read_sql(q, pconn)
df

Ma_sinh_vien         Ho_ten Ma_mon  Diem
0           C0355    Văn Thị Mai   CSD1   9.0
1           C0355    Văn Thị Mai   TTNT   9.5
2           C0355    Văn Thị Mai   TOA2   9.5
3           C0355    Văn Thị Mai   TOA1   9.5
4           C0355    Văn Thị Mai   NMLT   7.5
...           ...            ...    ...   ...
8726        V0900  Vương Phú Thu   LQL2   5.5
8727        V0900  Vương Phú Thu   LQL1   5.0
8728        V0900  Vương Phú Thu   KTLT   7.0
8729        V0900  Vương Phú Thu   CTDL   7.0
8730        V0900  Vương Phú Thu   CSD2   8.0

[8731 rows x 4 columns]

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_sinh_vien</th>
      <th>Ho_ten</th>
      <th>Ma_mon</th>
      <th>Diem</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>C0092</td>
      <td>Trần Thảo Nhung</td>
      <td>LQL2</td>
      <td>5.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>C0004</td>
      <td>Đinh Thị Thanh Dung</td>
      <td>CSD1</td>
      <td>5.5</td>
    </tr>
    <tr>
      <th>2</th>
      <td>C0004</td>
      <td>Đinh Thị Thanh Dung</td>
      <td>CSD2</td>
      <td>9.5</td>
    </tr>
    <tr>
      <th>3</th>
      <td>C0004</td>
      <td>Đinh Thị Thanh Dung</td>
      <td>CTDL</td>
      <td>6.5</td>
    </tr>
    <tr>
      <th>4</th>
      <td>C0004</td>
      <td>Đinh Thị Thanh Dung</td>
      <td>KTLT</td>
      <td>6.0</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [65]:
# Câu 3c: Liệt kê danh sách các sinh viên (gồm các cột mã, họ tên, ngày sinh) có học môn ‘Cơ sở dữ liệu’
q = '''select s."Ma_sinh_vien", concat(concat(s."Ho", \' \'), s."Ten") as "Ho_ten", s."Ngay_sinh"
        from sinh_vien s
        where s."Ma_sinh_vien" in
        (select distinct kq."Ma_sinh_vien"
         from ket_qua_2 kq left join mon_hoc m on m."Ma_mon" = kq."Ma_mon"
         where m."Ten" like \'Cơ sở dữ liệu%\')
        '''
df = pd.read_sql(q, pconn)
df

Ma_sinh_vien              Ho_ten  Ngay_sinh
0          C0232      Phạm Nam Trọng 1999-12-15
1          C0028        Lâm Nam Sinh 1999-07-24
2          C0029    Khương Ngọc Minh 1999-10-28
3          C0030     Phạm Vĩnh Phong 1999-09-05
4          C0031        Lý Thảo Thúy 1999-07-05
..           ...                 ...        ...
868        V0896         Mã Nam Dũng 1999-11-01
869        V0897        Văn Nam Kiệt 1999-07-09
870        V0898       Bạch Nam Hiệu 1999-08-22
871        V0899  Đinh Văn Minh Tuấn 1999-07-09
872        V0900       Vương Phú Thu 1999-08-01

[873 rows x 3 columns]

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_sinh_vien</th>
      <th>Ho_ten</th>
      <th>Ngay_sinh</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>C0445</td>
      <td>Đàm Ngọc Thu</td>
      <td>1999-09-18</td>
    </tr>
    <tr>
      <th>1</th>
      <td>C0004</td>
      <td>Đinh Thị Thanh Dung</td>
      <td>1999-01-19</td>
    </tr>
    <tr>
      <th>2</th>
      <td>C0007</td>
      <td>Lê Thị Mai</td>
      <td>1999-09-13</td>
    </tr>
    <tr>
      <th>3</th>
      <td>C0009</td>
      <td>Thái Thị Thanh Nhung</td>
      <td>1999-07-18</td>
    </tr>
    <tr>
      <th>4</th>
      <td>C0011</td>
      <td>Hoàng Thảo Thúy</td>
      <td>1999-12-07</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [66]:
# Câu 3d: Liệt kê danh sách các sinh viên (gồm các cột mã, họ tên, ngày sinh) đạt điểm 10 trong bài thi
q = '''select s."Ma_sinh_vien", concat(concat(s."Ho", \' \'), s."Ten") as "Ho_ten", s."Ngay_sinh"
        from sinh_vien s
        where s."Ma_sinh_vien" in
        (select distinct kq."Ma_sinh_vien"
        from ket_qua_2 kq 
        where kq."Diem" = 10)
        '''
df = pd.read_sql(q, pconn)
df

Ma_sinh_vien               Ho_ten  Ngay_sinh
0          C0387         Lê Bạch Ngân 1999-04-29
1          C0030      Phạm Vĩnh Phong 1999-09-05
2          C0032            Mã Thị Ly 1999-05-30
3          C0037   Cao Văn Minh Phong 1999-08-12
4          C0038           Mã Bạch Mỹ 1999-07-18
..           ...                  ...        ...
496        V0892       Đặng Minh Minh 1999-06-29
497        V0893      Vương Thảo Dung 1999-04-14
498        V0894        Đặng Thảo Nga 1999-06-03
499        V0895  Thái Thị Thanh Trúc 1999-08-19
500        V0897         Văn Nam Kiệt 1999-07-09

[501 rows x 3 columns]

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_sinh_vien</th>
      <th>Ho_ten</th>
      <th>Ngay_sinh</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>V0620</td>
      <td>Lê Mạnh Bảo</td>
      <td>1999-02-24</td>
    </tr>
    <tr>
      <th>1</th>
      <td>C0007</td>
      <td>Lê Thị Mai</td>
      <td>1999-09-13</td>
    </tr>
    <tr>
      <th>2</th>
      <td>C0009</td>
      <td>Thái Thị Thanh Nhung</td>
      <td>1999-07-18</td>
    </tr>
    <tr>
      <th>3</th>
      <td>C0011</td>
      <td>Hoàng Thảo Thúy</td>
      <td>1999-12-07</td>
    </tr>
    <tr>
      <th>4</th>
      <td>C0017</td>
      <td>Lê Thị Thảo Đào</td>
      <td>1999-12-27</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [69]:
# Câu 3e: Liệt kê danh sách các sinh viên (gồm các cột mã, họ tên, ngày sinh) có điểm thi môn ‘Toán cao cấp’ nhỏ hơn 5 
q = '''select s."Ma_sinh_vien", concat(concat(s."Ho", \' \'), s."Ten") as "Ho_ten", s."Ngay_sinh"
        from sinh_vien s
        where s."Ma_sinh_vien" in
        (select distinct kq."Ma_sinh_vien"
        from ket_qua_2 kq left join mon_hoc m on m."Ma_mon" = kq."Ma_mon"
        where kq."Diem" < 5
        and m."Ten" like \'%Toán cao cấp%\')
        '''
df = pd.read_sql(q, pconn)
df

Ma_sinh_vien              Ho_ten  Ngay_sinh
0          C0028        Lâm Nam Sinh 1999-07-24
1          C0031        Lý Thảo Thúy 1999-07-05
2          C0041       Bạch Văn Hiệu 1999-10-06
3          C0056       Thái Thảo Đào 1999-09-19
4          C0058  Trần Văn Minh Tuấn 1999-05-30
..           ...                 ...        ...
141        V0875  Lâm Thị Thảo Trang 1999-07-14
142        V0880     Phạm Mạnh Trọng 1999-05-13
143        V0883   Vương Minh Phương 1999-06-21
144        V0885  Lâm Thị Thảo Hương 1999-04-02
145        V0900       Vương Phú Thu 1999-08-01

[146 rows x 3 columns]

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_sinh_vien</th>
      <th>Ho_ten</th>
      <th>Ngay_sinh</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>C0442</td>
      <td>Văn Thị Thanh Nga</td>
      <td>1999-08-09</td>
    </tr>
    <tr>
      <th>1</th>
      <td>C0004</td>
      <td>Đinh Thị Thanh Dung</td>
      <td>1999-01-19</td>
    </tr>
    <tr>
      <th>2</th>
      <td>C0007</td>
      <td>Lê Thị Mai</td>
      <td>1999-09-13</td>
    </tr>
    <tr>
      <th>3</th>
      <td>C0009</td>
      <td>Thái Thị Thanh Nhung</td>
      <td>1999-07-18</td>
    </tr>
    <tr>
      <th>4</th>
      <td>C0011</td>
      <td>Hoàng Thảo Thúy</td>
      <td>1999-12-07</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [72]:
# Câu 3f: Cho biết danh sách các môn học chưa có sinh viên đăng ký 
q = '''select *
        from mon_hoc
        where "Ma_mon" not in
        (select distinct "Ma_mon"
        from ket_qua_2
        where "Ma_mon" is not null)
        '''
df = pd.read_sql(q, pconn)
df

Empty DataFrame
Columns: [Ma_mon, Ten]
Index: []

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_mon</th>
      <th>Ten</th>
    </tr>
  </thead>
  <tbody>
  </tbody>
</table>
</div>
  
</details>

In [83]:
# Câu 3g: Cho biết danh sách các sinh viên (gồm các cột mã, họ tên, ngày sinh) chưa đăng ký học môn học nào
q = '''select s."Ma_sinh_vien"
        , concat(concat(s."Ho",' '),s."Ten") as "Ho_ten"
        , s."Ngay_sinh"
        from sinh_vien s
        where s."Ma_sinh_vien" not in
        (select  distinct k."Ma_sinh_vien" from ket_qua k where k."Ma_sinh_vien" is not null)
'''
df = pd.read_sql(q, pconn)
df

Empty DataFrame
Columns: [Ma_sinh_vien, Ho_ten, Ngay_sinh]
Index: []

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_sinh_vien</th>
      <th>Ho_ten</th>
      <th>Ngay_sinh</th>
    </tr>
  </thead>
  <tbody>
  </tbody>
</table>
</div>
  
</details>

In [ ]:
# Câu 4: Sử dụng ibm_db API, truy vấn có nhóm

In [73]:
# Câu 4a: Thống kê tổng số sinh viên theo khoa. Có sắp tăng theo tổng số sinh viên
q = '''select "Ma_khoa", count(*) as "So_luong"
        from sinh_vien
        group by "Ma_khoa"
        '''
df = pd.read_sql(q, pconn)
df

Ma_khoa  So_luong
0      CN     500.0
1      TO     100.0
2      VL     400.0

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_khoa</th>
      <th>Tong_so_sinh_vien</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>TO</td>
      <td>100.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>CN</td>
      <td>235.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>VL</td>
      <td>400.0</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [74]:
# Câu 4b: Thống kê tổng học bổng theo khoa. Chỉ hiển thị những khoa có tổng học bổng >=3,000,000
q = '''select s."Ma_khoa", k."Ten", sum(s."Hoc_bong") as "Tong_hoc_bong"
        from sinh_vien s join khoa k on s."Ma_khoa"=k."Ma_khoa"
        group by s."Ma_khoa", k."Ten"
        having sum(s."Hoc_bong") >= 150000000
        '''
df = pd.read_sql(q, pconn)
df

Ma_khoa                       Ten  Tong_hoc_bong
0      CN  Khoa Công nghệ Thông tin      711000000
1      VL               Khoa Vật lý      616000000

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_khoa</th>
      <th>Ten</th>
      <th>Tong_hoc_bong</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>CN</td>
      <td>Khoa Công nghệ Thông tin</td>
      <td>397325000</td>
    </tr>
    <tr>
      <th>1</th>
      <td>TO</td>
      <td>Khoa Toán</td>
      <td>140000000</td>
    </tr>
    <tr>
      <th>2</th>
      <td>VL</td>
      <td>Khoa Vật lý</td>
      <td>616000000</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [75]:
# Câu 4c: Hiển thị danh sách sinh viên gồm mã, họ tên và điểm trung bình các môn học mà sinh viên có đăng ký học
q = '''select k."Ma_sinh_vien", concat(concat(s."Ho",' '),s."Ten") as "Ho_ten", avg(k."Diem") as "Diem_trung_binh"
        from ket_qua k join sinh_vien s on s."Ma_sinh_vien"=k."Ma_sinh_vien"
        group by k."Ma_sinh_vien", concat(concat(s."Ho",' '),s."Ten")
'''
df = pd.read_sql(q, pconn)
df

Ma_sinh_vien               Ho_ten  Diem_trung_binh
0          C0001     Khương Thảo Loan             6.50
1          C0002       Đặng Bạch Ngọc             7.20
2          C0003  Phạm Văn Minh Thiện             7.70
3          C0004  Đinh Thị Thanh Dung             6.90
4          C0005      Trần Mạnh Thiện             7.70
..           ...                  ...              ...
995        V0996      Trần Mạnh Thăng             6.25
996        V0997      Trần Vĩnh Thiện             7.50
997        V0998     Khương Thảo Hạnh             8.15
998        V0999       Đinh Ngọc Ngọc             7.30
999        V1000          Lý Mạnh Bảo             7.55

[1000 rows x 3 columns]

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_sinh_vien</th>
      <th>Ho_ten</th>
      <th>3</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>C0004</td>
      <td>Đinh Thị Thanh Dung</td>
      <td>6.950000</td>
    </tr>
    <tr>
      <th>1</th>
      <td>C0007</td>
      <td>Lê Thị Mai</td>
      <td>7.700000</td>
    </tr>
    <tr>
      <th>2</th>
      <td>C0009</td>
      <td>Thái Thị Thanh Nhung</td>
      <td>7.400000</td>
    </tr>
    <tr>
      <th>3</th>
      <td>C0011</td>
      <td>Hoàng Thảo Thúy</td>
      <td>7.944444</td>
    </tr>
    <tr>
      <th>4</th>
      <td>C0013</td>
      <td>Nguyễn Minh Thu</td>
      <td>8.000000</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [76]:
# Câu 4d: Hiển thị danh sách khoa gồm mã khoa, tên khoa và điểm trung bình của tất cả các sinh viên trong khoa
q = '''select k2."Ma_khoa"
        , k2."Ten"
        , avg(k."Diem") as "Diem_trung_binh"
        from ket_qua k 
        join sinh_vien s on s."Ma_sinh_vien"=k."Ma_sinh_vien"
        join khoa k2 on k2."Ma_khoa"=s."Ma_khoa"
        group by k2."Ma_khoa" , k2."Ten"
'''
df = pd.read_sql(q, pconn)
df

Ma_khoa                       Ten  Diem_trung_binh
0      CN  Khoa Công nghệ Thông tin           7.2188
1      TO                 Khoa Toán           7.2990
2      VL               Khoa Vật lý           7.1875

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_khoa</th>
      <th>Ten</th>
      <th>Diem_trung_binh</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>CN</td>
      <td>Khoa Công nghệ Thông tin</td>
      <td>7.544465</td>
    </tr>
    <tr>
      <th>1</th>
      <td>TO</td>
      <td>Khoa Toán</td>
      <td>7.551948</td>
    </tr>
    <tr>
      <th>2</th>
      <td>VL</td>
      <td>Khoa Vật lý</td>
      <td>7.474493</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [77]:
# Câu 4e: Hiển thị danh sách sinh viên gồm mã, họ tên, điểm trung bình và kết quả xếp loại học tập của sinh viên, 
# trong đó kết quả xếp loại học tập được xét dựa trên điểm trung bình theo quy tắc sau: 
#                                Điểm trung bình < 5 :  Yếu
#                         5 <= Điểm trung bình < 6.5 :  Trung bình
#                         6.5 <= Điểm trung bình < 8 :  Khá
#                               Điểm trung bình >= 8 :  Giỏi

q = '''select k."Ma_sinh_vien"
        , concat(concat(s."Ho",' '),s."Ten") as "Ho_ten"
        , avg(k."Diem") as "Diem_trung_binh"
        , case 
            when avg(k."Diem") < 5 then 'Yeu'
            when avg(k."Diem") < 6.5 then 'Trung binh'
            when avg(k."Diem") < 8 then 'Kha'
            else 'Gioi'
            end as "Xep_loai"
        from ket_qua k
        left join sinh_vien s on s."Ma_sinh_vien" = k."Ma_sinh_vien"
        group by k."Ma_sinh_vien"
        , concat(concat(s."Ho",' '),s."Ten")
'''
df = pd.read_sql(q, pconn)
df

Ma_sinh_vien               Ho_ten  Diem_trung_binh    Xep_loai
0          C0001     Khương Thảo Loan             6.50         Kha
1          C0002       Đặng Bạch Ngọc             7.20         Kha
2          C0003  Phạm Văn Minh Thiện             7.70         Kha
3          C0004  Đinh Thị Thanh Dung             6.90         Kha
4          C0005      Trần Mạnh Thiện             7.70         Kha
..           ...                  ...              ...         ...
995        V0996      Trần Mạnh Thăng             6.25  Trung binh
996        V0997      Trần Vĩnh Thiện             7.50         Kha
997        V0998     Khương Thảo Hạnh             8.15        Gioi
998        V0999       Đinh Ngọc Ngọc             7.30         Kha
999        V1000          Lý Mạnh Bảo             7.55         Kha

[1000 rows x 4 columns]

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_sinh_vien</th>
      <th>Ho_ten</th>
      <th>Diem_trung_binh</th>
      <th>Xep_loai</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>C0004</td>
      <td>Đinh Thị Thanh Dung</td>
      <td>6.950000</td>
      <td>Khá</td>
    </tr>
    <tr>
      <th>1</th>
      <td>C0007</td>
      <td>Lê Thị Mai</td>
      <td>7.700000</td>
      <td>Khá</td>
    </tr>
    <tr>
      <th>2</th>
      <td>C0009</td>
      <td>Thái Thị Thanh Nhung</td>
      <td>7.400000</td>
      <td>Khá</td>
    </tr>
    <tr>
      <th>3</th>
      <td>C0011</td>
      <td>Hoàng Thảo Thúy</td>
      <td>7.944444</td>
      <td>Khá</td>
    </tr>
    <tr>
      <th>4</th>
      <td>C0013</td>
      <td>Nguyễn Minh Thu</td>
      <td>8.000000</td>
      <td>Giỏi</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [78]:
# Câu 4f: Hiển thị danh sách các môn học cùng với điểm thi cao nhất mà sinh viên đã đạt được trong môn học đó, 
# kết quả trả về gồm mã môn, tên môn, điểm thi cao nhất của sinh viên 
q = '''select k."Ma_mon"
        ,m."Ten"
        , max(k."Diem") as "Diem_cao_nhat"
        from ket_qua k
        left join mon_hoc m on m."Ma_mon"=k."Ma_mon"
        group by k."Ma_mon", m."Ten"
'''
df = pd.read_sql(q, pconn)
df

Ma_mon                  Ten  Diem_cao_nhat
0   CSD1      Cơ sở dữ liệu 1           10.0
1   CSD2      Cơ sở dữ liệu 2           10.0
2   CTDL     Cấu trúc dữ liệu           10.0
3   KTLT   Kỹ thuật Lập trình           10.0
4   LQL1  Lập trình Quản lý 1           10.0
5   LQL2  Lập trình Quản lý 2           10.0
6   NMLT   Nhập môn Lập trình           10.0
7   TOA1      Toán cao cấp A1           10.0
8   TOA2      Toán cao cấp A2           10.0
9   TTNT     Trí tuệ nhân tạo           10.0

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_mon</th>
      <th>Ten</th>
      <th>Diem_thi_cao_nhat</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>CSD1</td>
      <td>Cơ sở dữ liệu 1</td>
      <td>10.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>CSD2</td>
      <td>Cơ sở dữ liệu 2</td>
      <td>10.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>CTDL</td>
      <td>Cấu trúc dữ liệu</td>
      <td>10.0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>KTLT</td>
      <td>Kỹ thuật Lập trình</td>
      <td>10.0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>LQL1</td>
      <td>Lập trình Quản lý 1</td>
      <td>10.0</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [80]:
# Câu 4g: Hiển thị danh sách các môn học cùng với số lượng sinh viên đã đăng ký học các môn học đó, 
# kết quả trả về gồm mã môn, tên môn, số lượng sinh viên đã đăng ký 
q = '''select k."Ma_mon"
        , m."Ten"
        , count(*) as "So_sinh_vien"
        from ket_qua_2 k join mon_hoc m on m."Ma_mon"=k."Ma_mon"
        group by  k."Ma_mon", m."Ten"
'''
df = pd.read_sql(q, pconn)
df

Ma_mon                  Ten  So_sinh_vien
0   CSD1      Cơ sở dữ liệu 1         873.0
1   CSD2      Cơ sở dữ liệu 2         873.0
2   CTDL     Cấu trúc dữ liệu         873.0
3   KTLT   Kỹ thuật Lập trình         873.0
4   LQL1  Lập trình Quản lý 1         873.0
5   LQL2  Lập trình Quản lý 2         873.0
6   NMLT   Nhập môn Lập trình         873.0
7   TOA1      Toán cao cấp A1         873.0
8   TOA2      Toán cao cấp A2         873.0
9   TTNT     Trí tuệ nhân tạo         874.0

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_mon</th>
      <th>Ten</th>
      <th>So_luong_sinh_vien</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>CSD1</td>
      <td>Cơ sở dữ liệu 1</td>
      <td>930.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>CSD2</td>
      <td>Cơ sở dữ liệu 2</td>
      <td>921.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>CTDL</td>
      <td>Cấu trúc dữ liệu</td>
      <td>907.0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>KTLT</td>
      <td>Kỹ thuật Lập trình</td>
      <td>919.0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>LQL1</td>
      <td>Lập trình Quản lý 1</td>
      <td>928.0</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [81]:
# Câu 4h: Hiển thị danh sách các môn học có số lượng sinh viên đăng ký học đông nhất, 
# kết quả trả về gồm mã môn, tên môn và số lượng sinh viên đăng ký 
q = '''select t."Ma_mon", m."Ten", t."So_sinh_vien"
        from(
        select k."Ma_mon", count(*) as "So_sinh_vien", rank() over(order by count(*) desc) as "Thu_tu"
        from ket_qua_2 k
        group by  k."Ma_mon"
        ) t
        left join mon_hoc m on m."Ma_mon" = t."Ma_mon"
        where t."Thu_tu" = 1
'''
df = pd.read_sql(q, pconn)
df

Ma_mon               Ten  So_sinh_vien
0   TTNT  Trí tuệ nhân tạo         874.0

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_mon</th>
      <th>Ten</th>
      <th>So_luong_sinh_vien</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>CSD1</td>
      <td>Cơ sở dữ liệu 1</td>
      <td>930.0</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [82]:
# Câu 4i: Thống kê số lượng sinh viên đậu và rớt của từng môn, biết rằng sinh viên rớt khi điểm thi nhỏ hơn 5, 
# kết quả trả về gồm mã môn, tên môn, số sinh viên đậu, số sinh viên rớt 
q = '''select k."Ma_mon", m."Ten"
        , sum(case when k."Diem" < 5 then 1 else 0 end) as "Rot" 
        , sum(case when k."Diem" >= 5 then 1 else 0 end) as "Dau" 
        from ket_qua_2 k left join mon_hoc m on m."Ma_mon"=k."Ma_mon"
        group by k."Ma_mon", m."Ten"
'''
df = pd.read_sql(q, pconn)
df

Ma_mon                  Ten  Rot  Dau
0   CSD1      Cơ sở dữ liệu 1   60  813
1   CSD2      Cơ sở dữ liệu 2   72  801
2   CTDL     Cấu trúc dữ liệu   81  792
3   KTLT   Kỹ thuật Lập trình   67  806
4   LQL1  Lập trình Quản lý 1   62  811
5   LQL2  Lập trình Quản lý 2   72  801
6   NMLT   Nhập môn Lập trình   78  795
7   TOA1      Toán cao cấp A1   76  797
8   TOA2      Toán cao cấp A2   77  796
9   TTNT     Trí tuệ nhân tạo   75  799

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_mon</th>
      <th>Ten</th>
      <th>Pass</th>
      <th>Fail</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>CSD1</td>
      <td>Cơ sở dữ liệu 1</td>
      <td>930</td>
      <td>0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>CSD2</td>
      <td>Cơ sở dữ liệu 2</td>
      <td>921</td>
      <td>0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>CTDL</td>
      <td>Cấu trúc dữ liệu</td>
      <td>907</td>
      <td>0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>KTLT</td>
      <td>Kỹ thuật Lập trình</td>
      <td>919</td>
      <td>0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>LQL1</td>
      <td>Lập trình Quản lý 1</td>
      <td>928</td>
      <td>0</td>
    </tr>
    <tr>
      <th>5</th>
      <td>LQL2</td>
      <td>Lập trình Quản lý 2</td>
      <td>920</td>
      <td>0</td>
    </tr>
    <tr>
      <th>6</th>
      <td>NMLT</td>
      <td>Nhập môn Lập trình</td>
      <td>912</td>
      <td>0</td>
    </tr>
    <tr>
      <th>7</th>
      <td>TOA1</td>
      <td>Toán cao cấp A1</td>
      <td>911</td>
      <td>0</td>
    </tr>
    <tr>
      <th>8</th>
      <td>TOA2</td>
      <td>Toán cao cấp A2</td>
      <td>911</td>
      <td>0</td>
    </tr>
    <tr>
      <th>9</th>
      <td>TTNT</td>
      <td>Trí tuệ nhân tạo</td>
      <td>914</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [ ]:
conn.close()